textile/chat-features.textile

```diff
@@ -9,5 +9,148 @@ jump_to:
+
+There are four room lifecycle operations: @ATTACH@, @DETACH@, @RELEASE@ and @RETRY@. @RETRY@ is an internal-only operation. In order to ensure the integrity of the chat room status, these operations are carried out atomically.
+
+h4(#rooms-lifecycle-operations). Room Lifecycle Operations
+
+* @(CHA-RL1)@ A room must be explicitly @ATTACHED@ to.
+** @(CHA-RL1a)@ @[Testable]@ If the room is already in the @ATTACHED@ status, then this operation is no-op.
+** @(CHA-RL1b)@ @[Testable]@ If the room is in the @RELEASING@ status, the operation shall be rejected with error code @102102@.
+** @(CHA-RL1c)@ @[Testable]@ If the room is in the @RELEASED@ status, the operation shall be rejected with error code @102103@.
+** @(CHA-RL1d)@ @[Testable]@ If a room lifecycle operation is already in progress, this operation shall wait until the current operation completes before continuing.
+** @(CHA-RL1e)@ @[Testable]@ Notwithstanding the above points, when the attachment operation begins, the room shall be transitioned to the @ATTACHING@ status.
+** @(CHA-RL1f)@ @[Testable]@ The underlying @contributors@ will have their Realtime Channels attached in sequence.
+** @(CHA-RL1g)@ When all @contributors@ Realtime Channels successfully attach, the operation is now complete and the room is considered attached.
+*** @(CHA-RL1g1)@ @[Testable]@ The room status shall be transitioned to @ATTACHED@.
+*** @(CHA-RL1g2)@ @[Testable]@ Any contributors to room that have a pending discontinuity event against them, must be notified of this fact, using the error that caused the original discontinuity.
+*** @(CHA-RL1g3)@ @[Testable]@ Any transient disconnect timeouts shall be cleared.
+** @(CHA-RL1h)@ If a one of the @contributors@ Realtime Channels fails to attach, then the operation has failed and must be rolled back.
+*** @(CHA-RL1h1)@ @[Testable]@ The @attach@ call must throw an @ErrorInfo@. The code for this error is determined by the feature that failed (see the @ErrorCodes@ enum in the JS SDK for more information).
+*** @(CHA-RL1h2)@ @[Testable]@ If the underlying Realtime Channel entered the @SUSPENDED@ state, then the room status will transition to @SUSPENDED@, using the Realtime Channel error as the @cause@ field of the @ErrorInfo@.
+*** @(CHA-RL1h3)@ @[Testable]@ When the room enters the @SUSPENDED@ status, the room lifecycle shall enter the recovery loop (TODO: REF) and attempt to re-establish attachment. The attach operation shall fail using the @ErrorInfo@ associated with the @SUSPENDED@ room status as the error.
```

Comment:
> the room lifecycle shall enter the recovery loop (TODO: REF)

Does this refer to the CHA-RL5 `RETRY` operation? If so, what is the process for one operation to initiate another? e.g. what if there are other operations waiting to be performed per e.g. CHA-RL1d?

Reply:
Yes - I've fixed the reference.

Have tried to clarify :)

Reply:
Thanks! It's still not really clear to me what "asynchronously" means in the updated text though. Does it mean that it should be executed after all other pending operations have completed? I’m wondering whether this spec point and CHA-RL1d could be more clearly expressed in terms of a queue of operations, which get picked off one after another?

Reply:
Have attempted to make it clearer!